File: LeetCode_Marathons/LCM1/Hard/Reducing_Dishes/a.cpp
class Solution {
public:
    vector<int> globalSatisfaction;
    int getSum(int count) {
        int n = globalSatisfaction.size();
        int counter = 1, ans = 0;
        for (int i = n - count; i <= n - 1; i++) { 
            ans += counter * globalSatisfaction[i];
            counter++;
        }
        return ans;
    }


    int maxSatisfaction(vector<int>& satisfaction) {
        int n = satisfaction.size();
        sort(satisfaction.begin(), satisfaction.end());
        globalSatisfaction = satisfaction;

        int l = 0, r = n, mid;
        while (l < r) {
            mid = (l + r + 1) / 2;
            if (getSum(mid) > getSum(mid - 1)) l = mid;
            else r = mid - 1;
        }

        return getSum(l);
    }
};

File: LeetCode_Marathons/LCM1/Hard/Reducing_Dishes/a.py
class Solution:
    globalSatisfaction = []
    
    def getSum(self, count: int) -> int:
        n = len(globalSatisfaction)
        ans, counter = 0, 1

        for i in range(n - count, n):
            ans += counter * globalSatisfaction[i]
            counter += 1

        return ans

    def maxSatisfaction(self, satisfaction: List[int]) -> int:
        global globalSatisfaction


        n = len(satisfaction)
        satisfaction.sort()
        globalSatisfaction = satisfaction
        l, r, mid = 0, n, 0
        while l < r:
            mid = (l + r + 1) // 2
            if self.getSum(mid) > self.getSum(mid - 1): l = mid
            else: r = mid - 1
        
        return self.getSum(l)